File: runtimes/python/src/engineering_ui_capabilities_runtime/core/secrets.py
"""Secret reference and resolver protocols (§15.1).

Contracts store references only (e.g. an environment-variable name or a
project-defined secret-provider key), never values. Resolution happens at
the latest practical point and returns a redaction-aware wrapper so a
resolved value can never be accidentally serialized, logged, or printed.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol, runtime_checkable


@dataclass(frozen=True)
class SecretReference:
    """A reference to a secret. Carries no value."""

    provider: str
    key: str

    def __repr__(self) -> str:  # pragma: no cover - trivial
        return f"SecretReference(provider={self.provider!r}, key={self.key!r})"


class RedactedSecret:
    """Wraps a resolved secret value. `str()`/`repr()` always redact; the
    raw value is only reachable via the explicit `reveal()` call, which
    callers must not pass to a logger, exception, or serialized outcome.
    """

    __slots__ = ("_value",)

    def __init__(self, value: str) -> None:
        self._value = value

    def reveal(self) -> str:
        return self._value

    def __repr__(self) -> str:
        return "RedactedSecret(***)"

    def __str__(self) -> str:
        return "***"

    def __eq__(self, other: object) -> bool:
        return isinstance(other, RedactedSecret) and self._value == other._value

    def __hash__(self) -> int:  # pragma: no cover - trivial
        return hash(("RedactedSecret", self._value))


@runtime_checkable
class SecretResolver(Protocol):
    """Resolves a `SecretReference` to a `RedactedSecret` at the latest
    practical point. Implementations MUST NOT log the resolved value.
    """

    def resolve(self, ref: SecretReference) -> RedactedSecret: ...

File: packages/capabilities-runtime-ts/src/secrets.ts
/**
 * Secret-reference protocol (§15.1). Contracts and configuration store
 * references such as an environment-variable name or a project-defined
 * secret-provider key — never the secret value itself. Resolution happens
 * at the latest practical point via a `SecretResolver`, and resolved
 * secrets are redaction-aware: `toString()`/JSON serialization never
 * exposes the raw value, only an explicit `.reveal()` does.
 */

export interface SecretReference {
  readonly kind: 'secret-reference'
  /** Provider-defined lookup key, e.g. an environment-variable name. */
  readonly ref: string
  /** Optional named provider; defaults to the resolver's own default provider. */
  readonly provider?: string
}

export function secretReference(ref: string, provider?: string): SecretReference {
  return provider === undefined ? { kind: 'secret-reference', ref } : { kind: 'secret-reference', ref, provider }
}

export function isSecretReference(value: unknown): value is SecretReference {
  return (
    typeof value === 'object' &&
    value !== null &&
    (value as { kind?: unknown }).kind === 'secret-reference' &&
    typeof (value as { ref?: unknown }).ref === 'string'
  )
}

const REDACTED = '[REDACTED]'

/**
 * A resolved secret. Never store the raw value on a plain field: the raw
 * value is only reachable via `.reveal()`, and every other access path
 * (`toString`, `toJSON`, default `util.inspect`/console formatting via
 * `[Symbol.for('nodejs.util.inspect.custom')]`) is redacted.
 */
export class ResolvedSecret {
  readonly reference: SecretReference
  private readonly rawValue: string

  constructor(reference: SecretReference, rawValue: string) {
    this.reference = reference
    this.rawValue = rawValue
  }

  /** Deliberately reveals the raw secret value. Callers must not log or serialize the result. */
  reveal(): string {
    return this.rawValue
  }

  toString(): string {
    return REDACTED
  }

  toJSON(): string {
    return REDACTED
  }
}

export interface SecretResolver {
  resolve(reference: SecretReference): ResolvedSecret | Promise<ResolvedSecret>
}
